### PR TITLE
Add deprecation_list and index_image_resolved attributes [CLOUDDST-3836]

### PR DESCRIPTION
**Pull request by pbortlov**
* add deprecation_list  and index_image_resolved attributes due to iiblib-3.0.0 changes

https://projects.engineering.redhat.com/browse/CLOUDDST-3836

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -49,10 +49,12 @@ def read_content(filepath):
     "Programming Language :: Python :: Implementation :: PyPy",
 ]
 
+
 def get_requirements():
     with open("requirements.txt") as f:
         return f.read().splitlines()
 
+
```
Comment (pbortlov) [inline setup.py:57]: added by tox

File: tests/test_iib_ops.py [hunk 2]
```diff
@@ -213,6 +213,7 @@ def add_bundles_mock_calls_tester(
         binary_image="binary-image",
         overwrite_from_index=True,
         overwrite_from_index_token="overwrite_from_index_token",
+        deprecation_list=["bundle1"],
```
Comment (pbortlov) [inline tests/test_iib_ops.py:216]: Should I create a new function similar to `def add_bundles_mock_calls_tester` where I specify `deprecation_list` argument? I would do that but it seems ther is already a lot of duplication between `def add_bundles_mock_calls_tester` and `add_bundles_mock_calls_tester_not_called`

**Comment by pbortlov** [comment]
@midnightercz @lipoja @querti PTAL

File: setup.py (hunk 1 above)
Comment (yashvardhannanavati) [reply setup.py:57]: I don't think this project uses Python black.

File: README.rst [hunk 3]
```diff
@@ -43,6 +43,7 @@ Push the created index image to Pulp
     --bundle container-registry.example.com/bundle/image:123
     --arch x86_64
     --skip-quay
+    --deprecation-list ["container-registry.example.com/index/bundle-image:latest"]
```
Comment (yashvardhannanavati) [inline README.rst:46]: Should be something like ` --deprecation-list container-registry.example.com/index/bundle-image:latest, container-registry.example.com/repo/bundle-image:2`..The `[.. ]` won'r work with the command line. It will have to be a comma separated str

File: tests/test_iib_ops.py [hunk 4]
```diff
@@ -293,7 +294,7 @@ def remove_operators_mock_calls_tester(
     "extra_args,push_items,mock_calls_tester",
     [
         (
-            [],
+            ["--deprecation-list", ["bundle1"]],
```
Comment (yashvardhannanavati) [inline tests/test_iib_ops.py:297]: Shouldn't this be `"--deprecation-list", "bundle1"`

Comment (yashvardhannanavati) [reply tests/test_iib_ops.py:297]: Could you also add a test where there are more than one bundles in the deprecation_list?

**Review by yashvardhannanavati — CHANGES_REQUESTED** [review]

File: setup.py (hunk 1 above)
Comment (pbortlov) [reply setup.py:57]: it does. it is in tox.ini 
```[testenv:pycodestyle]
deps=
    {[py]deps}
    pycodestyle
    black
basepython = python3.7```

File: tests/test_iib_ops.py (hunk 2 above)
Comment (midnightercz) [reply tests/test_iib_ops.py:216]: I think you can enhance the existing test

File: setup.py (hunk 1 above)
Comment (yashvardhannanavati) [reply setup.py:57]: Ah, I missed that. Sorry

File: README.rst (hunk 3 above)
Comment (pbortlov) [reply README.rst:46]: good point

File: pubtools/iib/iib_ops.py [hunk 5]
```diff
@@ -248,6 +248,12 @@
     "required": False,
     "type": str,
 }
+ADD_CMD_ARGS[("--deprecation-list",)] = {
+    "group": "IIB service",
+    "help": "List of deprecated bundles",
```
Comment (chandwanitulsi) [inline pubtools/iib/iib_ops.py:253]: specifying comma separated list could be useful in help message

File: tests/fake_task_manager.py [hunk 6]
```diff
@@ -31,6 +31,7 @@ def setup_task(
             "binary_image": binary_image,
             "binary_image_resolved": binary_image + "-resolved",
             "index_image": "index_image",
+            "index_image_resolved": "index_image_resolved",
```
Comment (chandwanitulsi) [inline tests/fake_task_manager.py:34]: Could you help me in understanding if adding index_image_resolved is directly related to adding deprecation_list or was it needed to be added anyway?

Comment (pbortlov) [reply tests/fake_task_manager.py:34]: I found it was not implemented there. I don't remember if it failed without that or not.. I will try to check if it can be removed.

Comment (pbortlov) [reply tests/fake_task_manager.py]: removed

**Review by querti — APPROVED** [review]

**Review by chandwanitulsi — APPROVED** [review]

File: tests/utils.py [hunk 7]
```diff
@@ -21,6 +21,7 @@ def setup_task(
         overwrite_from_index_token=None,
         state_seq=("in_progress", "finished"),
         op_type="add",
+        deprecation_list=None,
```
Comment (yashvardhannanavati) [inline tests/utils.py:24]: Should we default this to an empty list instead?

File: tests/test_iib_ops.py [hunk 8]
```diff
@@ -293,7 +294,7 @@ def remove_operators_mock_calls_tester(
     "extra_args,push_items,mock_calls_tester",
     [
         (
-            [],
+            ["--deprecation-list", "bundle1,bundle2"],
```
Comment (yashvardhannanavati) [inline tests/test_iib_ops.py:297]: Could you add a parameter here testing when deprecation-list is passed as an empty string?

File: tests/utils.py [hunk 9]
```diff
@@ -35,7 +36,9 @@ def setup_task(
             "binary_image": binary_image,
             "binary_image_resolved": binary_image + "-resolved",
             "bundle_mapping": {},
+            "deprecation_list": deprecation_list,
             "index_image": "feed.com/index/image:tag",
+            "index_image_resolved": "feed.com/index/image:tag",
```
Comment (yashvardhannanavati) [inline tests/utils.py:41]: I don't see this related to this particular change. Could you please add this as a separate commit?

Also, it would be nice if you could squash your previous two commits into one.

File: tests/utils.py (hunk 7 above)
Comment (pbortlov) [reply tests/utils.py:24]: No, it is not a good practice adding an empty list to the function parameters.
Anti-pattern: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

File: tests/utils.py (hunk 9 above)
Comment (pbortlov) [reply tests/utils.py:41]: I will remove that.

File: tests/test_iib_ops.py (hunk 4 above)
Comment (pbortlov) [reply tests/test_iib_ops.py]: added

File: tests/utils.py (hunk 7 above)
Comment (pbortlov) [reply tests/utils.py:24]: I set the variable in the lines below to the []

File: tests/test_iib_ops.py (hunk 8 above)
Comment (pbortlov) [reply tests/test_iib_ops.py:321]: added and change is made in source code too not only at tests

File: tests/test_iib_ops.py [hunk 10]
```diff
@@ -213,6 +213,56 @@ def add_bundles_mock_calls_tester(
...
+    )
+    fixture_iib_client.assert_called_once_with(
+        "iib-server", auth=fixture_iib_krb_auth.return_value, ssl_verify=False
+    )
+    fixture_pulplib_repo_sync.assert_called_once()
+    assert fixture_pulplib_repo_sync.mock_calls[0].args[0].feed == "https://feed.com"
+    fixture_pulplib_repo_publish.assert_called_once()
+
+
+def add_bundles_mock_calls_tester_empty_deprecation_list(
```
Comment (yashvardhannanavati) [inline tests/test_iib_ops.py:226]: Optional: You can probably use `pytest.parameterize` for this and the test below.

**Review by yashvardhannanavati — APPROVED** [review]

**Comment by yashvardhannanavati** [comment]
@pbortlov The CI seems to be failing. Could you please fix that before merging this?

**Comment by pbortlov** [comment]
> @pbortlov The CI seems to be failing. Could you please fix that before merging this?

#21 @yashvardhannanavati

**Comment by pbortlov** [comment]
@midnightercz @querti @yashvardhannanavati @chandwanitulsi @lipoja I add another commit to fix index_image_resolved.

File: tests/utils.py (hunk 7 above)
Comment (lipoja) [inline tests/utils.py:24]: I am just wondering why does it accepts the parameter when it is not used.
I know that there are others unused parameters as well. But do we really want to be empty every time?

should not line 49  look like this ` "deprecation_list": [] if not deprecation_list else deprecation_list`?

File: tests/utils.py [hunk 11]
```diff
@@ -37,6 +37,7 @@ def setup_task(
             "binary_image_resolved": binary_image + "-resolved",
             "bundle_mapping": {},
             "index_image": "feed.com/index/image:tag",
+            "index_image_resolved": "feed.com/index/image-resolved:tag",
```
Comment (lipoja) [inline tests/utils.py:40]: would not be better to use some fake URL? or at least fake domain? (just because feed.com is legitimate site)

File: tests/utils.py (hunk 7 above)
Comment (pbortlov) [reply tests/utils.py:24]: solved

File: tests/utils.py (hunk 11 above)
Comment (pbortlov) [reply tests/utils.py]: changed

File: tests/test_iib_ops.py (hunk 10 above)
Comment (pbortlov) [reply tests/test_iib_ops.py:226]: I would have to research how to use parametrize of this function in a test that uses parametrize too.

File: pubtools/iib/iib_ops.py [hunk 12]
```diff
@@ -248,6 +248,12 @@
     "required": False,
     "type": str,
 }
+ADD_CMD_ARGS[("--deprecation-list",)] = {
+    "group": "IIB service",
+    "help": "Comma separated list of deprecated bundles",
+    "required": False,
+    "type": str,
```
Comment (yashvardhannanavati) [inline pubtools/iib/iib_ops.py:255]: I have seen that if I do not submit a `deprecation_list` in pub, it submits `--deprecation-list` without anything after it. Should we have a default value `" "` here to make sure that is handled?

File: tests/test_iib_ops.py [hunk 13]
```diff
@@ -293,10 +343,20 @@ def remove_operators_mock_calls_tester(
...
+            ["--deprecation-list", "bundle1,bundle2"],
+            [operator_1_push_item_pending, operator_1_push_item_pushed],
+            add_bundles_mock_calls_tester_deprecation_bundles,
+        ),
+        (
+            ["--deprecation-list", "bundle1"],
             [operator_1_push_item_pending, operator_1_push_item_pushed],
             add_bundles_mock_calls_tester,
         ),
+        (
```
Comment (yashvardhannanavati) [inline tests/test_iib_ops.py:355]: Can I ask what happens if `--deprecation-list` is `None`

File: pubtools/iib/iib_ops.py (hunk 12 above)
Comment (yashvardhannanavati) [reply pubtools/iib/iib_ops.py:255]: Example
```
[DEBUG   ] pubtools-iib-add-bundles --arch ppc64le --bundle \
[DEBUG   ]   bundle --deprecation-list \
[DEBUG   ]   --index-image quay.io/foo/index-image:v4.6 --pulp-url https://pulp-url \
[DEBUG   ]   --pulp-user admin --skip-pulp --iib-krb-ktfile /home/user/ktfile --iib-krb-principal user@EXAMPLE \
[DEBUG   ]   --overwrite-from-index --iib-server iib-server
```

Comment (pbortlov) [reply pubtools/iib/iib_ops.py:255]: Line #538 should take care of it

File: tests/test_iib_ops.py (hunk 13 above)
Comment (pbortlov) [reply tests/test_iib_ops.py:355]: Line #538 in iib_ops.py. It will not take an argument.

**Review by midnightercz — APPROVED** [review]

File: pubtools/iib/iib_ops.py (hunk 12 above)
Comment (yashvardhannanavati) [reply pubtools/iib/iib_ops.py:255]:
```
 ➜  pubtools-iib git:(paya_pr) ✗ pubtools-iib-add-bundles --arch ppc64le --bundle  bundle --deprecation-list --index-image quay.io/foo/index-image:v4.6         
usage: pubtools-iib-add-bundles [-h] [--pulp-url PULP_URL] [--pulp-user PULP_USER]
                                [--pulp-password PULP_PASSWORD] [--pulp-insecure]
                                [--pulp-repository PULP_REPOSITORY] [--iib-insecure] --iib-server IIB_SERVER
                                --iib-krb-principal IIB_KRB_PRINCIPAL [--iib-krb-ktfile IIB_KRB_KTFILE]
                                [--iib-cnr-token IIB_CNR_TOKEN] [--index-image INDEX_IMAGE]
                                [--binary-image BINARY_IMAGE] [--arch ARCH] [--overwrite-from-index]
                                [--overwrite-from-index-token OVERWRITE_FROM_INDEX_TOKEN] [--skip-pulp]
                                [--skip-quay] [--quay-dest-repo QUAY_DEST_REPO] [--quay-user QUAY_USER]
                                [--quay-password QUAY_PASSWORD] [--quay-remote-exec]
                                [--quay-ssh-remote-host QUAY_SSH_REMOTE_HOST]
                                [--quay-ssh-remote-host-port QUAY_SSH_REMOTE_HOST_PORT]
                                [--quay-ssh-reject-unknown-host] [--quay-ssh-username QUAY_SSH_USERNAME]
                                [--quay-ssh-password QUAY_SSH_PASSWORD]
                                [--quay-ssh-key-filename QUAY_SSH_KEY_FILENAME] [--quay-send-umb-msg]
                                [--quay-umb-url QUAY_UMB_URL] [--quay-umb-cert QUAY_UMB_CERT]
                                [--quay-umb-client-key QUAY_UMB_CLIENT_KEY]
                                [--quay-umb-ca-cert QUAY_UMB_CA_CERT] [--quay-umb-topic QUAY_UMB_TOPIC]
                                [--bundle BUNDLE] [--iib-legacy-org IIB_LEGACY_ORG]
                                [--deprecation-list DEPRECATION_LIST]
pubtools-iib-add-bundles: error: argument --deprecation-list: expected one argument
```

Comment (yashvardhannanavati) [reply pubtools/iib/iib_ops.py:255]: It gives me this with the current code

Comment (yashvardhannanavati) [reply pubtools/iib/iib_ops.py:255]: On second thoughts, I think this should be the correct behavior if you do not provide an empty string. I'll change Pub to never submit that arg if the deprecation_list is empty

File: tests/test_iib_ops.py (hunk 10 above)
Comment (yashvardhannanavati) [reply tests/test_iib_ops.py:226]: I don't understand. You can simply combine this test and `add_bundles_mock_calls_tester_deprecation_bundles` using an approach similar to https://github.com/release-engineering/pubtools-iib/pull/20/files/20668514dc69ca31cabd8f905907a2057d7cb3e9#diff-9c4a86c181a4176bddb1e77c859da5be358901042d39f308d1db4a1eaeb70923R342

File: tests/test_iib_ops.py (hunk 13 above)
Comment (yashvardhannanavati) [reply tests/test_iib_ops.py:355]: As in the output I pasted above, I don't think #538 will work. `None` would essentially mean not even an empty string and simply the argument. So it'll falter even before it reaches #538

File: tests/test_iib_ops.py (hunk 10 above)
Comment (lipoja) [reply tests/test_iib_ops.py:226]: Yeah, I overlook that this is not test but just  some helper function. Right now I am not sure if can use parameterize (I would say we do not)

Comment (pbortlov) [reply tests/test_iib_ops.py:226]: `add_bundles_mock_calls_tester_empty_deprecation_list` is a mock function which we use bellow at `test_add_bundles_cli` line #399.

Comment (yashvardhannanavati) [reply tests/test_iib_ops.py:226]: oh! I missed that too. +1 to keep it separate
